AK: Implement quickselect Algorithm and use it for `AK::Statistics::median()`

Popaulol opened the pull request:
This PR includes an implementation of the quickselect Algorithm to find the kth smallest Element of a collection and uses that to improve  `AK::Statistics::median()`.
It also includes Testing both for `AK::quick_select` and `AK::Statistics`.
Additionally I've added some error and bounds checking to `AK::Statistics` as i said I would do in #16870 .

Since this my first larger scale Contribution to serenity i am open to a lot feedback, although I won't have that much time in the following week, so it might take me a few days to respond to/implement feedback.

Because I didn't wanna put my full name into the License Header for the files i created I put a generic SerenityOS developers there, i hope this is okay.

I also noticed a few more issues both in statistics and there are still some open Things in quick_select so i added relevant FIXMEs and TODOs there. So sadly this PR does increase in the Fixme count :-(

Popaulol commented:
This should fix the commit messages

Popaulol commented:
Finally, I thought it was the Header line that was limited to 72 chars not the entire commit message...

Does anyone know that fuzzer error is about?
I don't understand what the issue is there by looking at the log...

Popaulol commented:
Oh, these are build errors that apperently my local build doesn't give me?

That's interesting...

Popaulol commented:
This rebase should've fixed the ci build.

konradekk commented on AK/Statistics.h:
```diff
@@ -27,10 +27,23 @@ class Statistics {
     T const min() const
     {
+        // Lets Rather fail than read over the end of a collection
+        VERIFY(size() != 0);
```
most of numerals are linearly ordered; maybe we could define a `minimal` one (i.e. lowest than any other numeral) and use it just like zero is used in the `average`? (there it is just neutral element of sum, which could be defined as `zero`)

pretty much the same for maximum below? 🤔

Popaulol replied:
I've thought about using something like INT_MAX for min and INT_MIN for max (since those would be also useful starting values for the algorithm instead of using the first values). But Since we are generic, I'd need to find a way to find the upper and lower bounds of a template type which I wasn't able to find anywhere in the serenity codebase (only searched for like 5 mins though) so i just put a verify there for now.
If you know a good way to get the bounds of a templated types, that would be useful to fix this.

konradekk replied:
it’s pretty generic so it has potential to be used all over the place—it’ll be better to check with maintainers how they see that…

Popaulol replied:
Yeah, I think I leave it like this for now, crashing at a verify with a comment is better than crashing from a vector out of bounds or returning something unexpected.

konradekk commented on AK/QuickSelect.h:
```diff
@@ -0,0 +1,171 @@
+
+// All of these quick select implementations return the `index` of the resulting element, after the algorithm has run, not the element itself!
+template<typename Collection>
+size_t quick_select(Collection& collection, size_t k)
```
itʼs pretty non-obvious (at least for me) that ‘selectingʼ from the collection modifies it. i feel that at least a comment is in place… (iʼd be the best to reflect the exact side effect in the name in my opinion).

Popaulol replied:
Makes sense, i didn't think about it.
I'll add it to the comment.

BenWiederhake replied:
Just as a heads-up, this seems still unresolved, is this right?

Popaulol replied:
i thought i did adress it?
Probably got lost somewhere along in the rebase, thanks for reminding me.

Popaulol replied:
Now looking at this again since I've already got the repo opened.

At least for me in the diff view of the whole PR, this is showing my new and updated comment already? So I don't know what's going on...

BenWiederhake replied:
Oh! Alright, I thought you wanted to rename it to `quickselect_inplace` or similar, following the tradition to indicate things that might be surprising in the name.

Popaulol replied:
Ankles sense and is a good idea. I'll rename all of the variants.

Popaulol replied:
Makes not Ankles... Wrote that on my phone with autocorrect...

BenWiederhake commented on AK/QuickSelect.h:
```diff
@@ -0,0 +1,172 @@
+namespace PivotFunctions {
+
+// Just use the first element of the range as the pivot
+// Mainly used to debug the quicksort algorithm
```
I think you mean quick_select here?

Popaulol replied:
Yes I did.

BenWiederhake commented on AK/QuickSelect.h:
```diff
@@ -0,0 +1,172 @@
+}
+
+// Implementation detail of median_of_medians
+// Whilst this looks quadratic in runtime, it always gets called with 5 or less elements so can be considered constant runtime
```
fewer*

Popaulol replied:
Makes sense

BenWiederhake commented on AK/QuickSelect.h:
```diff
@@ -0,0 +1,172 @@
+size_t partition5(Collection& collection, size_t left, size_t right, LessThan less_than)
+{
+    VERIFY((right - left) <= 5);
+    size_t i = left + 1;
+    while (i <= right) {
```
Is there a particular reason that these aren't just `for`-loops?

Popaulol replied:
I don't think there's a good reason, I should definitely change this.

BenWiederhake commented on AK/QuickSelect.h:
```diff
@@ -0,0 +1,172 @@
+/*
```
(GitHub still doesn't support commenting on the commit message, so I'll just write it here)

collection* (lowercase)
option*
time* complexity*
useful*

BenWiederhake commented on AK/Statistics.h:
```diff
@@ -51,16 +51,17 @@ class Statistics {
-    // FIXME: Implement a better algorithm
     T const median()
     {
         quick_sort(m_values);
```
Could it be that you forgot to delete this line?

Popaulol replied:
No, it's deleted in a later commit (72a0face671fb505538dfd83346f41c2d4efd9a0) which is also in this pr where the algorithm is actually changed.

BenWiederhake commented on AK/Statistics.h:
```diff
@@ -53,8 +69,10 @@ class Statistics {
 
     T const median()
     {
-        quick_sort(m_values);
```
Ah, here you delete this line. It's good enough for me, but if you want some practice with `git commit --fixup` and `git rebase -i`, this is your opportunity to learn.

Popaulol replied:
I'm deleting it in the commit where I actually replaced the median implementation.

Since that's essentially seperate from adding the algorithm in itself (since that can be used for other things than computing medians) I put them into seperate commits.

BenWiederhake commented on Tests/AK/TestStatistics.cpp:
```diff
@@ -0,0 +1,58 @@
+    EXPECT_APPROXIMATE(odd_number_elements.median(), 3.0);
+    EXPECT_APPROXIMATE(even_number_elements.median(), 3.5);
+
+    // The Expected values for Standard deviation and variance where calculated by my school issued scientific calculator
```
expected*
standard*
were*

BenWiederhake reviewed (changes requested):
Can you post a few measurements? For example, how does it compare to "just quick_sort the entire thing"? Your first commit correctly mentions "for really large datasets", and that's also what I remember from CS classes: the quick_select has huge constants hidden in that O(n). I'm curious how big they are in this case. Perhaps add a benchmark for quick_sort and a benchmark for quick_select, with similar input sizes, where quick_select has its time to shine?

My guess is that quick_select is significantly slower for sizes lower than 10k. If that turns out to be the case, it might be a good idea to make a case distinction in Statistics.h.

Also: Hi! I'm also German, and know how hard it was to get into the habit of English capitalization rules. Keep it up :)

Popaulol commented:
> Can you post a few measurements? For example, how does it compare to "just quick_sort the entire thing"? Your first commit correctly mentions "for really large datasets", and that's also what I remember from CS classes: the quick_select has huge constants hidden in that O(n). I'm curious how big they are in this case. Perhaps add a benchmark for quick_sort and a benchmark for quick_select, with similar input sizes, where quick_select has its time to shine?
> 
> My guess is that quick_select is significantly slower for sizes lower than 10k. If that turns out to be the case, it might be a good idea to make a case distinction in Statistics.h.
> 
> Also: Hi! I'm also German, and know how hard it was to get into the habit of English capitalization rules. Keep it up :)

Thanks for all the feedback, 

Basically the quick select algorithm is single pivot quick sort but only done partially, it only recourses on the partition that contains the element we are searching for.

With my implementation I've added the option to change the way it picks the pivot for the partitioning. When using something like picking an element, this results in an average runtime of O(n) but a worst case (if we're unlucky with the pivots) of O(n²).
One of the pivot functions I added was therefore the median of medians which uses mutual recursion to pick a really good pivot, ensuring that even the worst case for quick select is O(n).
This obviously comes with a lot of overhead, so I've decided to not have it be the default but instead it has to manually requested by the programmer.

I think I willdo some performance testing on it using the naive version with quick sort and the different pivoting methods and looking at when it starts to outperform using something like picking a random pivot, or just using quick sort itself.

I don't even get Capitalization correct in German most of the time so please keep correcting me 🙃

I'm currently writing all these responses from a Bus on my laptop, so I can't actually change anything right now without it being a huge pain.
I'll address your reviews probably in the next few days, however most likely neither today nor tomorrow.

Thank you for the feedback

~ Staubi

Popaulol commented:
I did have a bit of time right now so i addressed the small things, some performance benchmarks will be done either on Wednesday or Saturday. I'll start writing the benchmarking code now, but i most likely don't have enough time to finish it today.

BenWiederhake commented:
No worries, take your time, this is a volunteer project. Don't let it stress you out :^)

Popaulol commented:
So i did a bit of benchmarking, i generated the benchmarking starting data the same way the quick_sort tests do, which might've had an impact on the results.

Running Tests with lists up to 100_000 elements, the first_element and middle_element partitioning methods are always outclassed by factors of 20 or more even by the native median() implementation using quicksort, so they're are, atleast for the data I used basically useless, although they are valid usecases atleast for middle_element as outlined by my code comments. 

With median of medians and random element, the performance of median of median starts to overtake random_element at about 4500 elements in the list. When comparing this to just doing sorting, the best of the two is quicker than quicksort after about 200 elements for odd element count inputs and 350 elements for even element count inputs.
This is because for an even number of elements we have to run quick_select twice, to get the two middle elements.

I will change the median implementation in Statistics to reflect the results of this benchmark.

Popaulol commented:
Should i rather ammend the original commit for this or make a new one?

BenWiederhake commented:
Lovely! I was afraid that the median of medians algorithm would only start to be faster at counts like a hundred million or something. Turns out that fear didn't come true. Great!

Since I don't know how you intend to change the code, I don't have a preference.

I'm guessing it's probably something like `if (size() > THRESHOLD_MEDIAN_OF_MEDIANS) { /* do the median of medians thing */ } else { /* copy, sort and pick */ }` ? I feel like both options of committing are reasonable. I'd like to ask you to add tests that are big enough to hit the case for sure, because letting this go completely untested sounds like a bad idea. Perhaps throw in a `VERIFY(statistics_object.size() > Statistics::THRESHOLD_MEDIAN_OF_MEDIANS);` in the test, or something to make sure that this doesn't go wrong in the future. Of course, if I guessed wrong and you were thinking about something else, ignore this advice :D

Btw, about the even case: The current implementation is perfectly fine.
I'm curious though, perhaps you know the answer: Is a better approach even possible? We're looking for the "smallest element bigger-ish than the median", but that will be really hard to get right in the presence of equal elements. For example, in the "unsorted" list `{1, 3, 3, 3, 4, 5, 6, 7}`, the element after the "first" median-number is strictly bigger than it; on the other hand, in the "unsorted" list `{0, 1, 3, 3, 3, 3, 6, 7}` the element after the "first" median-number is equal. However, the number of elements strictly smaller is different in each case, perhaps something can be done with that?
My point is: A linear scan surely beats median-of-medians, so if we can somehow avoid a second invocation, that might halve the running time. After all, you said that the threshold goes from 200 to 350.

Then again, it's perfectly fine if we merge it as-is and I try to optimize it myself :D

Popaulol commented:
Thanks ;)

I will look at this tommorrow ;)

Popaulol commented:
I will implement all these thresholds for quickselect into the "default" quickselect call (no custom pivot/less than function) and the even/odd distinction will go into median. So that all callers who don't want to worry about this themselves won't have to look at this and can just call quickselect.

Popaulol commented:
> Lovely! I was afraid that the median of medians algorithm would only start to be faster at counts like a hundred million or something. Turns out that fear didn't come true. Great!
> 
> Since I don't know how you intend to change the code, I don't have a preference.
> 
> I'm guessing it's probably something like `if (size() > THRESHOLD_MEDIAN_OF_MEDIANS) { /* do the median of medians thing */ } else { /* copy, sort and pick */ }` ? I feel like both options of committing are reasonable. I'd like to ask you to add tests that are big enough to hit the case for sure, because letting this go completely untested sounds like a bad idea. Perhaps throw in a `VERIFY(statistics_object.size() > Statistics::THRESHOLD_MEDIAN_OF_MEDIANS);` in the test, or something to make sure that this doesn't go wrong in the future. Of course, if I guessed wrong and you were thinking about something else, ignore this advice :D
> 
> Btw, about the even case: The current implementation is perfectly fine. I'm curious though, perhaps you know the answer: Is a better approach even possible? We're looking for the "smallest element bigger-ish than the median", but that will be really hard to get right in the presence of equal elements. For example, in the "unsorted" list `{1, 3, 3, 3, 4, 5, 6, 7}`, the element after the "first" median-number is strictly bigger than it; on the other hand, in the "unsorted" list `{0, 1, 3, 3, 3, 3, 6, 7}` the element after the "first" median-number is equal. However, the number of elements strictly smaller is different in each case, perhaps something can be done with that? My point is: A linear scan surely beats median-of-medians, so if we can somehow avoid a second invocation, that might halve the running time. After all, you said that the threshold goes from 200 to 350.
> 
> Then again, it's perfectly fine if we merge it as-is and I try to optimize it myself :D

I don't think i will need to add Testing for the checking of the versions with quickselect, since all different pivoting methods already have their testing. But i will add Testing for the different paths median could take.

BenWiederhake commented on Tests/AK/TestStatistics.cpp:
```diff
@@ -12,6 +12,8 @@ TEST_CASE(Statistics)
     // Setup Test Data
     AK::Statistics<double> odd_number_elements;
     AK::Statistics<double> even_number_elements;
+    AK::Statistics<double> odd_number_elements_201;
+    AK::Statistics<double> even_number_elements_350;
```
Can this go into a separate test case? Ideally, add something like `EXPECT(201 > Statistics::EVEN_NAIVE_MEDIAN_CUTOFF); EXPECT(350 > Statistics::ODD_NAIVE_MEDIAN_CUTOFF);`, so that the test fails when those numbers change.

Also, I believe this would make the condition `(size() <= EVEN_NAIVE_MEDIAN_CUTOFF && size() % 2 == 0)` true, doesn't it? In that case it would still fail to cover the `quick_select` invocation.

Popaulol replied:
Makes sense. I will do this.

The quick select invocations will be covered by both of of those. I explicitly checked using debug prints i obiously didn't commit. The two 6 element versions will deal with the naive version whilst the two big ones will deal with both quickselect cases.

BenWiederhake replied:
Awesome! Can you do me a favor and walk me through this? I really don't understand what's going on then.

We initialize the dataset to have 350 elements. so `size()` will return 350. `EVEN_NAIVE_MEDIAN_CUTOFF` is defined in `Statistics.h` to be 350. So I would expect that the `(size() <= EVEN_NAIVE_MEDIAN_CUTOFF && size() % 2 == 0)` branch is taken. You claim that something different happens. You're probably right, but could you explain why?

Popaulol replied:
No I was actually thinking you meant that branch being taken, since that's what I am testing here.

The who cases with large datasets are supposed to be testing the even and odd quick select cases respectively.
The Two short ones with either 5 or 6 elements are meant to be testing the naive implementation of median.

I think I'm missunderstanding you're original question?

BenWiederhake replied:
Ah, I never spelled out the actual question I have, my bad, sorry.

Let's look at the invocation `even_number_elements_350.median()`.
- Do I read that right that the *intention* is that this covers the "large dataset, even quick select" case?
- With the above reasoning, I would strongly expect that the "SMALL dataset, even quickSORT" branch is taken.

I added a few prints, just to see where me logic goes wrong: https://github.com/BenWiederhake/serenity/commit/7bfbecf86c89e9e518214685f84dfc0d97bfd1e9

This is the output:
```
$ ./TestStatistics 
Running 1 cases out of 1.
Running test 'Statistics'.
Running TestStatistics median, small odd
Statistics::median(): size()=5, odd quickSORT taken
Running TestStatistics median, small even
Statistics::median(): size()=6, even quickSORT taken
Running TestStatistics median, large odd
Statistics::median(): size()=201, odd quickSELECT taken
Running TestStatistics median, large even
Statistics::median(): size()=350, even quickSORT taken
Finished TestStatistics median
Completed test 'Statistics' in 0ms
Finished 1 tests and 0 benchmarks in 0ms (0ms tests, 0ms benchmarks, 0ms other).
Out of 1 tests, 1 passed and 0 failed.
```

As you can see, the even quickSELECT branch is not covered by this test, even though it tries to do so. Let's improve that, by:
- changing "350" to "360",
- replacing the numbers in the variable names by "large", and
- adding a `VERIFY()` that checks that the hard-coded numbers definitely are larger than the respective threshold.

Popaulol replied:
Thank you, I'll do that once i got time.

BenWiederhake reviewed (changes requested):
Only a small fix in the test ^^
(And a reminder that a previous item isn't resolved yet.)

Popaulol commented:
Might take me a few days to find the time to actually do anything. But thanks for all the great feedback ;)

konradekk commented on AK/Statistics.h:
```diff
@@ -27,10 +31,23 @@ class Statistics {
+        if (size() == 0)
+            return 0;
+
+        return (float)sum() / size();
```
note: not sure about use cases etc., but shall we deal with `sum` going out of bounds in any way? (integer overflow?)

Popaulol replied:
There's a TODO about this already iirc?

Popaulol replied:
I checked there is one on line 28 and it was there already before this PR.

I thought about how i could fix it when i originally started on this, didn't find a good solution because of the genericness of the class. (How would you even check for overflow if you're dealing with different sizes of ints and floats based on templates.).
So if you have a good idea on how i could deal with it i'd add it to the PR but if not i think we should just leave it open for the future.

konradekk replied:
if you want, you might place a comment there (a `TODO`?) to signal, that weʼre aware of potential overflow problems while computing the sum. yet itʼs not, by all means, necessary!

Popaulol replied:
> if you want, you might place a comment there (a `TODO`?) to signal, that weʼre aware of potential overflow problems while computing the sum/average. yet itʼs not, by all means, necessary!

I think having two todos for exactly the same thing inside the same file is redundant and confusing. When it gets resolved someone might forget to remove the other todos etc.

konradekk replied:
not exactly the same for me… but by all means I wonʼt insist on that! 😅

(the thing is that `sum` here is some kind of a shortcut here; you might want to sum partially and reduce with division etc.; sum itself needs just data store big enough to hold the result — those are, of course, speculations so let us ignore them for now.)

resolving! 😃

Popaulol replied:
Okay, I'll add notes there when I have time.

Popaulol replied:
done

Popaulol commented:
During one of my rebases, i apperently somehow lost the seperate commit for the AK::Statistics Tests and it got somehow put in with the Threshhold commit...

Anyone have a good idea if, and if so how i can seperate these again?

konradekk commented on AK/Statistics.h:
```diff
@@ -51,16 +71,26 @@ class Statistics {
     }
 
-    // FIXME: Implement a better algorithm
     T const median()
```
same issue as with the `quickselect`; both `quickselect_inplace` and `quick_sort` modify the underlying structure. i believe that we should indicate that more clearly, esp. with apparently read-only operation like `median` computation.

Popaulol replied:
We can modify `m_values` freely since it's not directly visible to the users and it doesn't change the results of any other we have implemented in statistics right now.

Since the old Median implementation modified it aswell I think it's fine to modify it in place.

BenWiederhake reviewed (approved):
LGTM, all past comments are addressed. I somehow thought it was merged already :D

Popaulol commented:
> LGTM, all past comments are addressed. I somehow thought it was merged already :D

Good ;) Don't worry abt it ;)

Popaulol commented:
This actually also fixes a bug in the old Median solution I just noticed, since that one assumes the list would be one indexed, which it obviously isn't.

GMTA commented on AK/QuickSelect.h:
```diff
@@ -0,0 +1,169 @@
+static inline int random_int(size_t min, size_t max)
+{
+    return min + round_to<size_t>(random_double() * (max - min));
+}
```
Hi, author of the Tubes' variant here. We're returning an `int` but working with `size_t min` and `max`, that doesn't look right?

Popaulol replied:
Yeah, I think I had to mess with it to make type checking happy. I'll make it consistent in the evening today.
